File: fixtures/scale/scala/medium/src/calc/eval/Evaluator.scala
package calc.eval

import calc.ast._
import calc.parser.OperatorTable

final class Evaluator(environment: Environment, registry: FunctionRegistry) {

  def evaluate(expr: Expr): Double = expr match {
    case NumberExpr(value) => value
    case VariableExpr(name) => environment.lookup(name)
    case AssignExpr(name, value) =>
      val result = evaluate(value)
      environment.define(name, result)
      result
    case UnaryExpr(op, operand) =>
      val inner = evaluate(operand)
      if (op == '-') -inner else inner
    case BinaryExpr(op, left, right) =>
      OperatorTable.apply(op, evaluate(left), evaluate(right))
    case CallExpr(function, arguments) =>
      registry.invoke(function, arguments.map(evaluate).toArray)
  }
}

File: fixtures/scale/kotlin/medium/src/calc/eval/Evaluator.kt
package calc.eval

import calc.ast.AssignExpr
import calc.ast.BinaryExpr
import calc.ast.CallExpr
import calc.ast.Expr
import calc.ast.NumberExpr
import calc.ast.UnaryExpr
import calc.ast.VariableExpr
import calc.parser.applyOperator

class Evaluator(
    private val environment: Environment,
    private val registry: FunctionRegistry,
) {
    fun evaluate(expr: Expr): Double = when (expr) {
        is NumberExpr -> expr.value
        is VariableExpr -> environment.lookup(expr.name)
        is AssignExpr -> evaluate(expr.value).also { environment.define(expr.name, it) }
        is UnaryExpr -> if (expr.operator == '-') -evaluate(expr.operand) else evaluate(expr.operand)
        is BinaryExpr -> applyOperator(expr.operator, evaluate(expr.left), evaluate(expr.right))
        is CallExpr -> registry.invoke(expr.function, expr.arguments.map { evaluate(it) }.toDoubleArray())
    }
}
